(#3344) Adds ParameterBlock Handling to Chocolatey Script Runner

JPRuskin opened the pull request:
## Description Of Changes

This PR introduces the ability for ChocolateyInstall, ChocolateyBeforeModify, and ChocolateyUninstall scripts to have and utilise PowerShell `param()` blocks to handle parameters.

It does this by adding a function (`Get-PackageScriptParameters`), which calculates what parameters a script has and passes any parameters found by the existing function `Get-PackageParameters` that match the parameter names to the script. Optionally, we could handle aliases, but this implementation does not yet.

The new function is called within the ChocolateyScriptRunner before invoking the script, and the resultant hashtable is "splatted" with the call. When no matching parameters are found (on the script _and_ in packageparameters), this results in no change to behaviour.

It's worth noting that though this does allow for usage of parameter attributes and validation, such as `[Parameter(Mandatory)]`, `[ValidateSet()]` and the like, Chocolatey can currently swallow some useful output (e.g. with SomeString being a mandatory parameter):
![image](https://github.com/user-attachments/assets/52ceaa09-dc54-4c3f-ad48-5f89c1b32754)
vs the PowerShell default:
![image](https://github.com/user-attachments/assets/999c6da4-900a-446f-a508-6adb9b77c109)
This could possibly be improved in the future.

It adds the functionality, but doesn't remove the existing parameter handling - so all existing packages will continue working.

### SMA Update
As a part of this, because of the use of the `System.Management.Automation.Language` to assess the param block, the `System.Management.Automation` assembly present in the `lib/powershell` folder was updated from a version matching the version shipped with PowerShell 2.0 (best guess) to a version from our earliest currently supported version of Windows - 2012.

This DLL isn't shipped in any of our builds; it's just used to ensure our code uses methods et al that are supported by the specific DLL we have embedded in the repository.

Specifically, I grabbed this from the uncertain/fun provenance of a Windows Server 2012 VM, in the folder `C:\Windows\Microsoft.NET\assembly\GAC_MSIL\System.Management.Automation\v4.0_3.0.0.0__31bf3856ad364e35`. There's not a similar version on nuget.org, annoyingly, which only has versions from 2018 onwards starting at 6.0.4.

There is an argument that this is a breaking change, as it may no longer work on machines we don't support - but that "break" has already happened. We _do not_ support Windows versions prior to 2012 (and that only in Azure). Everything else is additive-only, and backwards compatible via a "lovely" new extension I'm looking to publish.

## Motivation and Context

This allows for standardisation of how package maintainers handle parameters within their scripts, and helps Chocolatey Community Repository users _and_ moderators to more easily understand what parameters can be used, what default values are used, and what (if any) validation is in place.

## Testing

- Grab https://chocolatey.org/install.ps1 and the modified version of the Chocolatey nupkg package, copy them to your test environment.
- `Set-ExecutionPolicy RemoteSigned Process -Force` and Install with `.\Install.ps1 -ChocolateyDownloadUrl .\Path\To\chocolatey.nupkg`
- Optional: install a random package without parameters to ensure I've not broken the existing functionality `choco install cloudflared --confirm`
- Optional: install a random package with parameters to ensure I've not broken the existing functionality `choco install Temurin21jre --params="/ADDLOCAL=FeatureMain,FeatureEnvironment,FeatureJarFileRunWith,FeatureJavaHome" --confirm`
- Grab a package with param blocks available, e.g. [testextension.1.0.0.zip](https://github.com/user-attachments/files/20596910/testextension.1.0.0.zip) (attached as a zip, as nupkg is not a supported file type) and copy it to the test environment (or have it in a repository)
- `choco install testextension --params='/SomeString:A String Value' --confirm` (some massaging required for source)
    ![image](https://github.com/user-attachments/assets/eefa152c-b990-4f30-865a-a73765d6cc47)

> NB: testextension will fail to install if the PowerShell parameter value "SomeString" provided doesn't match the package parameter found by `Get-PackageParameters`. The install script is partially shown below:

```powershell
param(
    [string]$SomeString
)

$PackageParameters = Get-PackageParameters

if ($PackageParameters.SomeString -ne $SomeString) {
    throw "'$SomeString' was not equal to '$($PackageParameters.SomeString)'"
}
```

### Operating Systems Testing

- Windows Server 2012
    ![image](https://github.com/user-attachments/assets/273d9f33-f0fa-45c3-b6e8-608d930959e7)
- Windows Server 2022
    ![image](https://github.com/user-attachments/assets/a6c48dc5-c4c4-45c2-a15d-5b5bf9e76061)
- Windows 11 (Sandbox)
    ![image](https://github.com/user-attachments/assets/eefa152c-b990-4f30-865a-a73765d6cc47)

I also have an extension package that adds the same functionality to any other extension-supporting version of Chocolatey CLI. This hopefully means that this can be an acceptable thing to use for Community packages.

![image](https://github.com/user-attachments/assets/647d737f-e6a3-4b4b-85f5-c285d69ea4b9)

I am very excited about this, despite the code being _utterly despicable_.

I have had to slightly revise my backwards compatibility happiness - it doesn't work with the earliest mention of Chocolatey extensions (0.9.10) because `Get-PackageParameters` doesn't exist at that point. I could fix this by packing a backup version fo the function in the extension, but I feel like supporting back to the 0.10.8 is enough for any unreasonable expectation:
![image](https://github.com/user-attachments/assets/a6ff3ffc-7d1c-4b10-af1a-eb6457f97859)

## Change Types Made

* ~[ ] Bug fix (non-breaking change).~
* [x] Feature / Enhancement (non-breaking change).
* ~[ ] Breaking change (fix or feature that could cause existing functionality to change).~ See SMA update.
* ~[ ] Documentation changes.~
* [x] PowerShell code changes.

## Change Checklist

* ~[ ] Requires a change to the documentation.~
* ~[ ] Documentation has been updated.~
* [x] Tests to cover my changes, have been added.
* [x] All new and existing tests passed?
* [x] PowerShell code changes: PowerShell v3 compatibility checked?

## Related Issue

Resolves #3344

JPRuskin committed:
(#3344) Adds Param Block Handling to chocolateyScriptRunner

Adds a basic function to the helper functions that parses a script file for parameters, and compares that list to available Package Parameters.

If there are matches, it returns a hashtable with matching names and their values.

This can then be used in the chocolateyScriptRunner to pass parameters into the param block.

The methods used all work back to PowerShell v3, though without any plan for backwards compatibility (e.g. considerations for using an extension package/module) it's hard to see this being that popular.

(#3344) Adds Tests for Get-ScriptParameters

Adding some tests to check the conversion from
PowerShell to C# cmdlet.

JPRuskin committed:
(#3344) Convert Get-PackageScriptParameter to helper cmdlet

It was pointed out that this PR should be targeted
at including new helper functions into the C# conversion
efforts.

This converts Get-PackageParameter to use a
helper class (shared between this and the new
cmdlet) and adds Get-PackageScriptParameters to the
Chocolatey.PowerShell compiled module.

Splits Get-functionality into private methods

Also refactors the helper method for package parameters to enable logging of sensitive parameter names (not values).

JPRuskin committed:
(#3344) Updates SMA DLL to Supported OS

Updates included SMA to 6.2.9200.24523.

This is now equivalent to a version shipped on Windows Server 2012
(the earliest OS Chocolatey products support).

This is necessary for the more advanced use of the AST parser,
and may allow us to benefit from upgrades elsewhere.

JPRuskin commented on src/Chocolatey.PowerShell/Helpers/PackageParameter.cs:
```diff
@@ -0,0 +1,93 @@
+                    paramStrings.Add(packageParameters);
+                }
+
+                // This should possibly only be implemented in the CLE codebase
+                var sensitivePackageParameters = EnvironmentHelper.GetVariable(
+                    cmdlet,
+                    "ChocolateyPackageParametersSensitive",
+                    EnvironmentVariableTarget.Process);
+                if (!string.IsNullOrEmpty(sensitivePackageParameters))
+                {
+                    logParams = false;
+                    paramStrings.Add(sensitivePackageParameters);
+                }
+            }
```
I suspect this should be removed, here, though I brought it over from the conversion branch so left it for discussion.

vexx32 replied:
I think it's fine to put here. The environment variable itself is set in CLE, and never in CLI, so this will essentially do nothing in CLI but IMO there is no harm doing it like this. IIRC `Start-ChocolateyProcessAsAdmin` also has a precedent for this kind of thing, exposing a parameter for `-SensitiveStatements` but it's never used in CLI and isn't supported except in CLE.

vexx32 committed:
(#3477) Add Get-PackageParameter helper cmdlet

Rewrite of the Get-PackageParameter(s) into C# cmdlet,
though there will be further changes made to refactor
it into a helper method in a later commit.

vexx32 commented on src/Chocolatey.PowerShell/Commands/GetPackageScriptParametersCommand.cs:
```diff
@@ -0,0 +1,69 @@
+
+        protected override void End()
+        {
+            var packageParameters = PackageParameter.GetParameters(this, Parameters);
+            var splatHash = new Hashtable(StringComparer.OrdinalIgnoreCase);
+
+            // Check what parameters the script has
+            Token[] tokensRef = null;
+            ParseError[] errorsRef = null;
+            var parsedAst = Parser.ParseFile(ScriptPath, out tokensRef, out errorsRef);
+            var scriptParameters = parsedAst.ParamBlock != null ? parsedAst.ParamBlock.Parameters.Select(p => p.Name.VariablePath.UserPath.ToString()).ToList() : new List<string>();
+            WriteVerbose($"Found {scriptParameters.Count()} parameter(s) in '{ScriptPath}'");
+
+            // For each of those in PackageParameters, add it to the splat
+            foreach (var parameter in scriptParameters)
+            {
+                if (packageParameters.ContainsKey(parameter))
+                {
+                    splatHash.Add(parameter, packageParameters[parameter]);
+                }
+            }
```
Is there a specific reason this wasn't pulled into a helper class/method as well?

JPRuskin replied:
Honestly, I figured that until we ever wanted to use it in a second place (which may end up being CLE) it made as much sense like this. I can do it, though, if preferred.

vexx32 commented on src/Chocolatey.PowerShell/Helpers/PackageParameter.cs:
```diff
@@ -0,0 +1,93 @@
+                        value = (object)true;
+                    }
+
+                    if (logParams)
+                    {
+                        cmdlet.WriteDebug($"Adding package param '{name}'='{value}'");
+                    }
```
I'm guessing this logic was kind of pre-existing but looking at this I _do_ wonder if we can split things out at some point to log the non-sensitive things and just skip logging sensitive data, maybe.

JPRuskin replied:
I was wondering the same! I'd happily modify it to do that.

vexx32 reviewed (commented):
I'll have to come back and review the tests added in a bit, but some initial comments. :3

JPRuskin committed:
(#3344) Removes Get-PackageParameter cmdlet, restores function scriptfile

As it was pointed out that removing a script from the helpers module was a breaking change, I have reverted that change.

The PackageParameter helper remains, used by PackageScriptParameters, but there are no longer any removals from the user contract.

JPRuskin commented:
I believe this is no longer a breaking change (as it doesn't remove or change any PowerShell helpers), makes all changes to helper functionality in C#, and doesn't break any business promises.

I suspect I've missed the boat on getting it merged into 2.5.0, though! Shocking.

JPRuskin committed:
(squash) Removes Unnecessary CSProj Components

The changes to use the newer csproj format in February (#816f85ba166433108517a3b90e37438860a49d0d) mean that the references in my previous commits are no longer required, and in fact conflict.

This commit removes those references, and should be squashed into the previous history.
